File: backend/routes/api_keys.py
from functools import wraps
from flask import Blueprint, request, jsonify, current_app
from models.user import User
from models.api_key import ApiKey, db
from routes.auth import get_current_user, jwt_required
import re
from datetime import datetime

bp = Blueprint('api_keys', __name__, url_prefix='/api/keys')

SUPPORTED_SERVICES = ['openai', 'claude', 'gemini', 'huggingface']

def validate_service_name(service_name):
    """Valide le nom du service"""
    return service_name.lower() in SUPPORTED_SERVICES

def validate_api_key_format(service_name, api_key):
    """Valide le format de la clé API selon le service"""
    if service_name == 'openai':
        return api_key.startswith('sk-') and len(api_key) > 20
    elif service_name == 'claude':
        return api_key.startswith('sk-ant-') and len(api_key) > 20
    elif service_name == 'gemini':
        return len(api_key) > 20  # Format générique
    elif service_name == 'huggingface':
        return api_key.startswith('hf_') and len(api_key) > 20
    return False

@bp.route('/', methods=['GET'])
@jwt_required
def list_api_keys():
    """Liste les clés API de l'utilisateur"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    keys = ApiKey.query.filter_by(user_id=user.id).all()
    return jsonify([key.to_dict() for key in keys]), 200

@bp.route('/', methods=['POST'])
@jwt_required
def create_api_key():
    """Crée une nouvelle clé API"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Données requises'}), 400
    
    service_name = data.get('service_name', '').lower()
    api_key = data.get('api_key', '')
    key_name = data.get('key_name', '')
    
    # Validations
    if not service_name or not api_key or not key_name:
        return jsonify({'error': 'service_name, api_key et key_name requis'}), 400
    
    if not validate_service_name(service_name):
        return jsonify({'error': f'Service non supporté. Services disponibles: {", ".join(SUPPORTED_SERVICES)}'}), 400
    
    if not validate_api_key_format(service_name, api_key):
        return jsonify({'error': 'Format de clé API invalide pour ce service'}), 400
    
    if len(key_name) < 3 or len(key_name) > 100:
        return jsonify({'error': 'Nom de clé entre 3 et 100 caractères'}), 400
    
    # Vérifier si une clé active existe déjà pour ce service
    existing_key = ApiKey.get_active_key_for_service(user.id, service_name)
    if existing_key:
        # Désactiver l'ancienne clé
        existing_key.is_active = False
    
    # Créer la nouvelle clé
    new_key = ApiKey(
        user_id=user.id,
        service_name=service_name,
        key_name=key_name
    )
    new_key.encrypt_key(api_key)
    
    try:
        db.session.add(new_key)
        db.session.commit()
        return jsonify(new_key.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erreur lors de la création'}), 500

@bp.route('/<int:key_id>', methods=['GET'])
@jwt_required
def get_api_key(key_id):
    """Récupère une clé API spécifique"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    api_key = ApiKey.query.filter_by(id=key_id, user_id=user.id).first()
    if not api_key:
        return jsonify({'error': 'Clé API non trouvée'}), 404
    
    return jsonify(api_key.to_dict()), 200

@bp.route('/<int:key_id>', methods=['PUT'])
@jwt_required
def update_api_key(key_id):
    """Met à jour une clé API"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    api_key = ApiKey.query.filter_by(id=key_id, user_id=user.id).first()
    if not api_key:
        return jsonify({'error': 'Clé API non trouvée'}), 404
    
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Données requises'}), 400
    
    # Seuls key_name et is_active peuvent être modifiés
    if 'key_name' in data:
        if len(data['key_name']) < 3 or len(data['key_name']) > 100:
            return jsonify({'error': 'Nom de clé entre 3 et 100 caractères'}), 400
        api_key.key_name = data['key_name']
    
    if 'is_active' in data:
        api_key.is_active = bool(data['is_active'])
    
    try:
        db.session.commit()
        return jsonify(api_key.to_dict()), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erreur lors de la mise à jour'}), 500

@bp.route('/<int:key_id>', methods=['DELETE'])
@jwt_required
def delete_api_key(key_id):
    """Supprime une clé API"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    api_key = ApiKey.query.filter_by(id=key_id, user_id=user.id).first()
    if not api_key:
        return jsonify({'error': 'Clé API non trouvée'}), 404
    
    try:
        db.session.delete(api_key)
        db.session.commit()
        return jsonify({'message': 'Clé API supprimée'}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erreur lors de la suppression'}), 500

@bp.route('/<int:key_id>/test', methods=['POST'])
@jwt_required
def test_api_key(key_id):
    """Teste une clé API"""
    user = get_current_user()
    if not user:
        return jsonify({'error': 'Utilisateur non trouvé'}), 404
    
    api_key = ApiKey.query.filter_by(id=key_id, user_id=user.id).first()
    if not api_key:
        return jsonify({'error': 'Clé API non trouvée'}), 404
    
    if not api_key.is_active:
        return jsonify({'error': 'Clé API désactivée'}), 400
    
    # Test de connexion (placeholder)
    try:
        success = api_key.test_connection()
        if success:
            api_key.last_used = datetime.utcnow()
            db.session.commit()
            return jsonify({'message': 'Clé API fonctionnelle', 'status': 'success'}), 200
        else:
            return jsonify({'message': 'Échec du test de connexion', 'status': 'error'}), 400
    except Exception as e:
        return jsonify({'error': 'Erreur lors du test'}), 500

@bp.route('/services', methods=['GET'])
@jwt_required
def get_supported_services():
    """Retourne la liste des services supportés"""
    return jsonify({'services': SUPPORTED_SERVICES}), 200

File: frontend/src/stores/apiKeys.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ApiKey, CreateApiKeyData, UpdateApiKeyData } from '@/types'
import { apiKeysApi } from '@/services/api'
import { useNotifications } from '@/composables/useNotifications'

export const useApiKeysStore = defineStore('apiKeys', () => {
  const apiKeys = ref<ApiKey[]>([])
  const supportedServices = ref<string[]>(['openai', 'claude', 'gemini', 'huggingface'])
  const isLoading = ref(false)
  const { showNotification } = useNotifications()

  // Actions
  const fetchApiKeys = async (): Promise<void> => {
    isLoading.value = true
    try {
      const response = await apiKeysApi.list()
      apiKeys.value = response.data
    } catch (error: any) {
      showNotification('Erreur lors du chargement des clés API', 'error')
      console.error('Fetch API keys error:', error)
    } finally {
      isLoading.value = false
    }
  }

  const createApiKey = async (data: CreateApiKeyData): Promise<boolean> => {
    isLoading.value = true
    try {
      const response = await apiKeysApi.create(data)
      apiKeys.value.push(response.data)
      showNotification(`Clé API "${data.key_name}" créée avec succès`, 'success')
      return true
    } catch (error: any) {
      const message = error.response?.data?.error || 'Erreur lors de la création de la clé API'
      showNotification(message, 'error')
      console.error('Create API key error:', error)
      return false
    } finally {
      isLoading.value = false
    }
  }

  const updateApiKey = async (id: number, data: UpdateApiKeyData): Promise<boolean> => {
    isLoading.value = true
    try {
      const response = await apiKeysApi.update(id, data)
      const index = apiKeys.value.findIndex(key => key.id === id)
      if (index !== -1) {
        apiKeys.value[index] = response.data
      }
      showNotification('Clé API mise à jour avec succès', 'success')
      return true
    } catch (error: any) {
      const message = error.response?.data?.error || 'Erreur lors de la mise à jour de la clé API'
      showNotification(message, 'error')
      console.error('Update API key error:', error)
      return false
    } finally {
      isLoading.value = false
    }
  }

  const deleteApiKey = async (id: number): Promise<boolean> => {
    isLoading.value = true
    try {
      await apiKeysApi.delete(id)
      apiKeys.value = apiKeys.value.filter(key => key.id !== id)
      showNotification('Clé API supprimée avec succès', 'success')
      return true
    } catch (error: any) {
      const message = error.response?.data?.error || 'Erreur lors de la suppression de la clé API'
      showNotification(message, 'error')
      console.error('Delete API key error:', error)
      return false
    } finally {
      isLoading.value = false
    }
  }

  const testApiKey = async (id: number): Promise<boolean> => {
    isLoading.value = true
    try {
      const response = await apiKeysApi.test(id)
      const result = response.data
      if (result.success) {
        showNotification('Test de la clé API réussi', 'success')
        return true
      } else {
        showNotification(`Échec du test: ${result.message}`, 'error')
        return false
      }
    } catch (error: any) {
      const message = error.response?.data?.error || 'Erreur lors du test de la clé API'
      showNotification(message, 'error')
      console.error('Test API key error:', error)
      return false
    } finally {
      isLoading.value = false
    }
  }

  const toggleApiKeyStatus = async (id: number): Promise<void> => {
    const apiKey = apiKeys.value.find(key => key.id === id)
    if (apiKey) {
      await updateApiKey(id, { is_active: !apiKey.is_active })
    }
  }

  const fetchSupportedServices = async (): Promise<void> => {
    try {
      const response = await apiKeysApi.getSupportedServices()
      supportedServices.value = response.data
    } catch (error) {
      console.error('Fetch supported services error:', error)
      // Utiliser les services par défaut en cas d'erreur
    }
  }

  // Getters
  const getApiKeysByService = (serviceName: string): ApiKey[] => {
    return apiKeys.value.filter(key => key.service_name === serviceName)
  }

  const getActiveApiKeys = (): ApiKey[] => {
    return apiKeys.value.filter(key => key.is_active)
  }

  const getApiKeyById = (id: number): ApiKey | undefined => {
    return apiKeys.value.find(key => key.id === id)
  }

  // Actions d'initialisation
  const init = async (): Promise<void> => {
    await Promise.all([
      fetchApiKeys(),
      fetchSupportedServices()
    ])
  }

  return {
    // State
    apiKeys,
    supportedServices,
    isLoading,
    
    // Actions
    fetchApiKeys,
    createApiKey,
    updateApiKey,
    deleteApiKey,
    testApiKey,
    toggleApiKeyStatus,
    fetchSupportedServices,
    init,
    
    // Getters
    getApiKeysByService,
    getActiveApiKeys,
    getApiKeyById
  }
})